### PR TITLE
llama : fix long/short rope factor selection by sequence length

### LINKED ISSUE #24823: Eval bug:Phi-3 / Phi-4 LongRoPE: short sequences silently use long-context RoPE factors when the allocated context exceeds original_max_position_embeddings
### Name and Version

llama-server version: 9608 (70b54e140) - the code driving this (get_rope_factors keying on n_ctx_seq in llama-model.cpp) is unchanged in current master, 9728 (fabde3b)

(same thing will happen in llama-cli, it's just easier to test in llama-server)

### Operating systems

Linux

### GGML backends

CPU, CUDA

### Hardware

It's independent of hardware, confirmed on multiple and the code cited below affects all.

### Models

phi3/4 family. It shows up most drastically in fine tunes but the bug is there for all of them, the default model just seems to be less affected by the issue but is equally wrong.

### Problem description & steps to reproduce

For LongRoPE models (Phi-3 / Phi-4 family), llama.cpp selects the long vs short RoPE factor set based on the allocated context size (n_ctx_seq), not the actual sequence length. As a result, when a model is loaded with a context larger than original_max_position_embeddings (4096 for Phi-4-mini), every sequence, including very short ones, is encoded with the long-context factors. The reference implementation (Hugging Face transformers) selects the factor set from the current sequence length, so short prompts get the short factors.

The divergence is silent, and its impact is model-dependent. Some models tolerate the mis-scaled encoding with no measurable change in quality, but it can be catastrophic for others, including a legitimate fine-tune that is correct at a smaller context and under the reference implementation. Because the impact varies by model, the failure is easily misattributed to the model or the fine-tuning process rather than to the serving path.

Environment

llama.cpp version: version: 9608 (70b54e140), reproduced with llama-server on CPU (-ngl 0), so the result is backend-independent. The faulty code is unchanged in current master (9728 / fabde3b); code references below.
Model: a Phi-4-mini-instruct fine-tune (Phi3 architecture), LongRoPE, rope.dimension_count = 96, rope.freq_base = 10000, rope.scaling.original_context_length = 4096, training context 131072. Served as an f16 GGUF with the LongRoPE factor tensors present, so this is not a quantization or missing-metadata issue.
Reference implementation: transformers 5.5.0.

Steps to reproduce

Minimal reproduction, native llama-server, CPU only, identical in every respect except the allocated context, which differs only by crossing the 4096 boundary:

```
# correct: allocation within the 4096 original_max -> short RoPE factors
llama-server -m phi4-mini-ft.f16.gguf -ngl 0 -c 4096 --host 127.0.0.1 --port 8099

# broken: same model, same prompt, same flags, only -c crosses 4096 -> long factors
llama-server -m phi4-mini-ft.f16.gguf -ngl 0 -c 8192 --host 127.0.0.1 --port 8099

# one greedy request in each case; the prompt is a single document of about 640 tokens:
curl -s localhost:8099/completion \
  -d '{"prompt":"<wrapped prompt>","n_predict":1024,"temperature":0,"cache_prompt":false}'
```

The server logs confirm the context took: n_ctx_seq (4096) and n_ctx_seq (8192) respectively, both well below the 131072 training context.

At -c 4096 the model returned a complete, correct extraction (field F1 = 1.000):

```
{"products": [{"product_id": "p1", "product_number": "WA-SL150", "product_name": "OmniLift Stacker SL-150"}],
 "options": [{"option_id": "o1", "product_id": "p1", "option_name": "Lift capacity"}, ...],
 "option_values": [{"value_id": "v1", "option_id": "o1", "value_name": "1000 kg"}, ...],
 "rules": [{"option_id": "o1", "min": 1, "max": 1}, ...]}
```

At -c 8192, with the same prompt, the model stopped grounding in the document after the first record and reproduces the output-format template with literal placeholder tokens (field F1 = 0.000):

```
{"products": [{"product_id": "p1", "product_number": "WA-SL150", "product_name": "OmniLift Stacker SL-150"}],
 "options": [{"option_id": "o1", "product_name": "option_name"}, {"option_id": "o2", "product_name": "option_name"}, ...],
 "option_values": [{"value_id": "v1", "option_id": "<ref>", "value_name": "value_name"}, ...],
 "rules": [{"option_id": "<ref>", "min": <int>, "max": <int or null>}, ...]}
```

The literal <ref>, value_name, <int or null>, and ... are placeholder tokens from the prompt's format description, not values read from the document. The only thing that changed between the two runs is -c.

Supporting measurements

I confirmed the issue direct via llama-server to rule out ollama (plus found the specific error in code) but I had first tested in ollama. Across the eight short documents that failed, served via Ollama 0.30.6 (bundled llama.cpp), greedy:

```
num_ctx = 4096    mean F1 = 0.992
num_ctx = 131072  mean F1 = 0.105
```

The same fine-tuned weights served through transformers 5.5.0 (which selects the short factors for these sub-4096 sequences) score about 0.94 on the full 105-document test set, matching the num_ctx <= 4096 llama.cpp result and confirming the weights are correct.

As a control, the base Phi-4-mini converted through the same pipeline is essentially unaffected by the allocation:

```
base    num_ctx = 4096    mean F1 = 0.913
base    num_ctx = 131072  mean F1 = 0.919
```

So the wrong factor selection is not catastrophic for every model. It is catastrophic for one that depends more heavily on precise positional encoding, which is exactly the kind of model that is correct at num_ctx <= 4096 and under the reference implementation. That model-dependence is what makes the issue easy to misattribute to the model rather than the serving path. tl;dr The issue is real but doesn't show up to a significant degree with the reference phi4 model, any finetune amplifies the existing bug though and becomes unusable quickly.

Expected behavior

A sequence whose length is at or below original_max_position_embeddings should be encoded with the short factor set, regardless of how much context capacity was allocated. This matches the LongRoPE reference and the transformers implementation.

In transformers (modeling_rope_utils._compute_longrope_parameters), long_factor is used only when seq_len > original_max_position_embeddings; otherwise short_factor is used.

Actual behavior

llama.cpp chooses the factor set from the allocated context, not the sequence length.

src/llama-model.cpp (get_rope_factors):

```cpp
ggml_tensor * llama_model::get_rope_factors(const llama_cparams & cparams, int il) const {
    const uint32_t n_ctx_seq = cparams.n_ctx_seq;

    // choose long/short freq factors based on the context size
    if (layers[il].rope_freqs != nullptr) {
        return layers[il].rope_freqs;
    }

    if (n_ctx_seq > hparams.n_ctx_orig_yarn) {
        return layers[il].rope_long;
    }

    return layers[il].rope_short;
}
```

n_ctx_seq is the allocated per-sequence context, set in src/llama-context.cpp:

```cpp
cparams.n_ctx_seq = cparams.n_ctx;                  // single sequence
// or
cparams.n_ctx_seq = cparams.n_ctx / cparams.n_seq_max;
```

So whenever n_ctx_seq > n_ctx_orig_yarn (4096), the long factors are returned for all positions, including a 700-token prompt that should use the short factors. For Phi-4-mini the long factors scale the inverse frequencies by up to roughly 47x on the low-frequency dimensions, which materially changes the positional encoding relative to the short factors the model (and the reference implementation) expect for short sequences.

Why this is easy to miss

The failure is silent. There is no warning, the JSON or text output is still well-formed, and for some inputs it is even correct, so the degradation looks like a model-quality or fine-tuning problem rather than a serving artifact. It only appears when the allocated context exceeds original_max_position_embeddings, which is common because tools default to or allow large contexts for a model whose training context is 131072.

Proposed fix

Choose the factor set from the actual sequence length (or the maximum position in the batch) rather than from the static allocation. This is what the reference implementation does, and for any sequence that stays within n_ctx_orig_yarn it is unambiguous and needs no special machinery: the short factors are simply correct, whatever the allocation. The current code returns the long factors for these sequences only because n_ctx_seq reflects the allocation rather than the length.

The one genuinely subtle case is a single sequence that crosses n_ctx_orig_yarn during generation while a KV cache is populated, since keys cached under one factor set cannot be cheaply re-roped to the other. That edge case is shared by any cached-attention implementation, including the reference, and it does not block fixing the common case where the whole sequence is at or below n_ctx_orig_yarn.

As a minimum mitigation, emit a warning when a LongRoPE model is loaded with n_ctx_seq > n_ctx_orig_yarn, so the silent degradation becomes visible.

Possible related issue

ggml-org/llama.cpp issue 12091 discusses Phi-4 RoPE handling. Worth checking for overlap or duplication before filing.

### First Bad Commit

not bisected, present in 9608 and unchanged in current master (9728). long standing, not a regression.

### Relevant log output

<details>
<summary>Logs</summary>

```console

```
</details>

### PR DESCRIPTION
**Pull request by o7si**
For LongRoPE models (Phi-3 / Phi-4 family), `get_rope_factors()` previously selected the long/short RoPE factors based on `cparams.n_ctx_seq`.

`n_ctx_seq` is the allocated context size, fixed at context creation and unrelated to the actual sequence length, so it is not a sound basis for choosing the factors — the selection should depend on the current sequence length.

This PR selects the factors based on the largest token position actually reached instead.

**Known remaining issues (left out to avoid a larger PR; I'm happy to extend this PR if needed)**

1. **No per-request factor selection.** A ubatch may mix short and long sequences, but the whole ubatch is selected with a single maximum position.
2. **History is not corrected when the context grows past the boundary.** Once a sequence grows beyond `original_max_position_embeddings`, keys already written to the KV-cache with the short factors are not re-rotated.

Fixes #24823

### CODE REVIEW

**Comment by o7si** [comment]
Verified with [`Phi-3.5-mini-instruct-Q4_K_M.gguf`](https://huggingface.co/bartowski/Phi-3.5-mini-instruct-GGUF/blob/main/Phi-3.5-mini-instruct-Q4_K_M.gguf). Same short prompt, greedy, CPU only; only `-c` differs:

```bash
llama-cli -m Phi-3.5-mini-instruct-Q4_K_M.gguf -p "List three benefits of regular exercise." -n 120 --temp 0 -ngl 0 -st -c <4096|8192>
```

On master the output changes when `-c` crosses 4096 (the short prompt wrongly gets the long factors at `-c 8192`). With this PR the output is identical at `-c 4096` and `-c 8192`, and matches master at `-c 4096`.

<details>
<summary>master, <code>-c 4096</code></summary>

```
1. Improved Cardiovascular Health: Regular exercise strengthens the heart muscle, improves blood circulation, and lowers blood pressure, which reduces the risk of heart disease and stroke.

2. Weight Management: Exercise helps in burning calories and maintaining a healthy weight by increasing metabolism and reducing body fat. This can also help prevent obesity-related health issues such as type 2 diabetes and certain types of cancer.

3. Mental Health Benefits: Regular physical activity has been
```

</details>

<details>
<summary>master, <code>-c 8192</code> — diverges</summary>

```
1. Improved cardiovascular health: Regular exercise strengthens the heart and improves circulation, which can reduce the risk of heart disease and stroke. It helps lower blood pressure, improve cholesterol levels, and enhance overall cardiovascular function, leading to a healthier and more efficient cardiovascular system.

2. Weight management: Exercise helps to burn calories and build muscle, which can aid in maintaining a healthy weight or losing excess weight. It also boosts metabolism, which
```

</details>

<details>
<summary>this PR, <code>-c 4096</code></summary>

```
1. Improved Cardiovascular Health: Regular exercise strengthens the heart muscle, improves blood circulation, and lowers blood pressure, which reduces the risk of heart disease and stroke.

2. Weight Management: Exercise helps in burning calories and maintaining a healthy weight by increasing metabolism and reducing body fat. This can also help prevent obesity-related health issues such as type 2 diabetes and certain types of cancer.

3. Mental Health Benefits: Regular physical activity has been
```

</details>

<details>
<summary>this PR, <code>-c 8192</code></summary>

```
1. Improved Cardiovascular Health: Regular exercise strengthens the heart muscle, improves blood circulation, and lowers blood pressure, which reduces the risk of heart disease and stroke.

2. Weight Management: Exercise helps in burning calories and maintaining a healthy weight by increasing metabolism and reducing body fat. This can also help prevent obesity-related health issues such as type 2 diabetes and certain types of cancer.

3. Mental Health Benefits: Regular physical activity has been
```

</details>

File: src/llama-batch.h [hunk 1]
```diff
@@ -27,6 +27,9 @@ struct llama_ubatch {
         return n_pos >= 3;
     }
 
+    // largest token position in the ubatch (max over all sequences)
+    int32_t pos_max() const;
```
Comment (ggerganov) [inline src/llama-batch.h:31]: This won't work correctly if there are multiple sequences in the batch and some are using short, other using long factors.

**Comment by ggerganov** [comment]
> History is not corrected when the context grows past the boundary. Once a sequence grows beyond original_max_position_embeddings, keys already written to the KV-cache with the short factors are not re-rotated.

Is this what the reference implementation does? I.e., when the sequence length exceeds the threshold, it rotates the keys in the cache?

**Comment by o7si** [comment]
> > History is not corrected when the context grows past the boundary. Once a sequence grows beyond original_max_position_embeddings, keys already written to the KV-cache with the short factors are not re-rotated.
> 
> Is this what the reference implementation does? I.e., when the sequence length exceeds the threshold, it rotates the keys in the cache?

No, it doesn't re-rotate the cache.

It's a Phi-specific workaround: `Phi3Model.prepare_inputs_for_generation` overrides the generic method and invalidates the whole KV cache when the sequence first crosses `original_max_position_embeddings`, then re-prefills with the long factors (only `phi3` / `phimoe` / `phi4_multimodal` do this):

https://github.com/huggingface/transformers/blob/0a2757da521a7a49b8143d9e0c938f08747d682e/src/transformers/models/phi3/modeling_phi3.py#L516-L528
```python
# Overwritten -- this model may need to switch between short and long rope, invalidating the cache in the
# process

# When the first time input length reached long and short factor switching point, enforce re-compute cache
# It will cause downside of slower at this single token position, however, better than current failure.
if (
    past_key_values
    and hasattr(self.config, "original_max_position_embeddings")
    and input_ids.shape[1] >= self.config.original_max_position_embeddings + 1
):
    past_length = past_key_values.get_seq_length()
    if past_length <= self.config.original_max_position_embeddings:
        past_key_values = None
```

HF discards the cache, but re-rotating the cached keys is likely more efficient and would give the same result.

Also, would it make sense to implement this generically instead of special-casing per model?

**Comment by ggerganov** [comment]
I think this LongRoPE design is very inefficient and I am doubtful that there is any point to support it precisely. Just falling back to the long factors as we do now for large `-c` is good for 99% of the cases.

Unless some new relevant model appears that really requires this mechanism for some reason, I don't think it's worth the extra logic. We can add some notes in the comments or print a warning in the logs.

**Comment by o7si** [comment]
Sounds good. For now a warning/log is the lightweight choice, so I'll go with that.

Worth noting the K-shift path already does this kind of re-rotation:

https://github.com/ggml-org/llama.cpp/blob/063d9c156e816ae3cf62db01f429a07a099afe97/src/llama-kv-cache.cpp#L1826-L1876

I'll benchmark it when I get the time, and if both the performance impact and the code change turn out to be small, I'll bring it up again for discussion.

File: src/llama-batch.h (hunk 1 above)
Comment (o7si) [reply src/llama-batch.h:31]: I looked into the HF code. I had assumed it selected the factors per-sequence, but it turns out it also selects them per-batch.

Specifically, the choice is driven by a single scalar taken over the whole batch, producing one `inv_freq` table for the entire forward:

https://github.com/huggingface/transformers/blob/0a2757da521a7a49b8143d9e0c938f08747d682e/src/transformers/modeling_rope_utils.py#L47-L80
```python
def longrope_frequency_update(self, position_ids, device, layer_type=None):
    """Longrope uses long factor if sequence is larger than original pretraining length, short otherwise."""
    seq_len = torch.max(position_ids) + 1
    ...
    if seq_len > original_max_position_embeddings:
        ...
        self.register_buffer(f"{prefix}inv_freq", long_inv_freq, persistent=False)
        ...
    else:
        ...
        self.register_buffer(f"{prefix}inv_freq", original_inv_freq, persistent=False)
        ...
```

So maybe we can drop the per-sequence implementation for now, since it would require fairly large changes.

**Comment by b-bargabus** [comment]
Thank you both! 

For what it's worth I think a warning for the edge case is sufficient. As it sat I went down a lot of wrong paths trying to diagnose the behavior I was seeing with Phi4-mini, this warning message would lead me to the root cause quickly so should be sufficient for the rare user that hits the edge case.

**Comment by ggerganov** [comment]
> Unless some new relevant model appears that really requires this mechanism for some reason, I don't think it's worth the extra logic. We can add some notes in the comments or print a warning in the logs.

I actually meant that we should keep the logic as it is on `master` (i.e. do not change `llama_model::get_rope_factors`) and simply print a warning upon context creation that long rope factors will be used for all inputs. The goal is to avoid the extra complexity from keeping track of the current max sequence position.

**Comment by b-bargabus** [comment]
> > Unless some new relevant model appears that really requires this mechanism for some reason, I don't think it's worth the extra logic. We can add some notes in the comments or print a warning in the logs.
> 
> I actually meant that we should keep the logic as it is on `master` (i.e. do not change `llama_model::get_rope_factors`) and simply print a warning upon context creation that long rope factors will be used for all inputs. The goal is to avoid the extra complexity from keeping track of the current max sequence position.

My opinion, without this fix (which is there now in the PR) it would probably be better to just remove phi support. It is very broken. In my case f1 collapsed from .9496 to .3494 across 105 document extraction tasks just by moving num_ctx from 4096 to 8192 under current master. The fix that o7si put in doesn't handle every case but it does significantly increase capabilities and brings it to par with the reference transformers implementation with a warning message about the much narrower case that will fail. I hope this is approved and merged.
Thank you,
BB.

**Comment by o7si** [comment]
> My opinion, without this fix (which is there now in the PR) it would probably be better to just remove phi support. It is very broken. In my case f1 collapsed from .9496 to .3494 across 105 document extraction tasks just by moving num_ctx from 4096 to 8192 under current master. The fix that o7si put in doesn't handle every case but it does significantly increase capabilities and brings it to par with the reference transformers implementation with a warning message about the much narrower case that will fail. I hope this is approved and merged.

@ggerganov Warning-only version is ready on [a separate branch](https://github.com/o7si/llama.cpp/commit/872bc76cdbd24dcc8d505f9e069e36d93f7d9dd7): `get_rope_factors` unchanged from master, just a warning at context creation. I haven't force-pushed it over this PR.

How would you like me to proceed?

**Comment by ggerganov** [comment]
@b-bargabus What is `f1`?

My expectation is that using only long rope factors should be pretty much equivalent to using short + long on generations beyond 4096 tokens - I don't see why there would be significant degradation.

And for generations with less than 4096 tokens, one can simply set `-c` such that the code selects the short factors.

**Comment by b-bargabus** [comment]
> @b-bargabus What is `f1`?
> 
> My expectation is that using only long rope factors should be pretty much equivalent to using short + long on generations beyond 4096 tokens - I don't see why there would be significant degradation.
> 
> And for generations with less than 4096 tokens, one can simply set `-c` such that the code selects the short factors.

f1 is the combined p and r metrics from evals, the standard for eval results. An f1 of 1.0 is perfect, 0.0 is absolute failure. Collapsing from .94 to .34 is a massive loss of intelligent results in objective testing.

I can understand how your expectations would seem intuitive but I have actual test results that show the actual behavior doesn't match that expectation. it is badly broken. Even the very simple test that o7si used failed badly on instruction following producing just two results when a clear prompt asked for three. Phi's rope settings are critical to the model behavior, using short on long or vice versa severely inhibits its abilities.

Setting -c in advance only works if you know the length of the input in advance and would fall apart in multiturn scenarios. What o7si just did in this PR is a significant improvement over master and makes phi usable except in a much more narrow edge case for which it warns. It also brings it to parity with the reference transformers implementation.

**Comment by o7si** [comment]
> Even the very simple test that o7si used failed badly on instruction following producing just two results when a clear prompt asked for three. 

Small clarification, that was just `-n 120` truncating the output. I re-ran it without the `-n` cap:

```
llama-cli -m Phi-3.5-mini-instruct-Q4_K_M.gguf -p "List three benefits of regular exercise." --temp 0 -ngl 0 -st -c <4096|8192>
```

<details>
<summary><code>-c 4096</code></summary>

```
1. Improved Cardiovascular Health: Regular exercise strengthens the heart muscle, improves blood circulation, and lowers blood pressure, which reduces the risk of heart disease and stroke.
2. Weight Management: Exercise helps in burning calories and maintaining a healthy weight by increasing metabolism and reducing body fat. This can also help prevent obesity-related health issues such as type 2 diabetes and certain types of cancer.
3. Mental Health Benefits: Regular physical activity has been shown to reduce symptoms of depression and anxiety, improve mood, and enhance overall mental well-being. Exercise stimulates the release of endorphins, which are natural mood-boosting chemicals in the brain. Additionally, it can improve sleep quality, increase self-esteem, and provide a sense of accomplishment.
```
</details>

<details>
<summary><code>-c 8192</code></summary>

```
1. Improved cardiovascular health: Regular exercise strengthens the heart and improves circulation, which can reduce the risk of heart disease and stroke. It helps lower blood pressure, improve cholesterol levels, and enhance overall cardiovascular function, leading to a healthier and more efficient cardiovascular system.
2. Weight management: Exercise helps to burn calories and build muscle, which can aid in maintaining a healthy weight or losing excess weight. It also boosts metabolism, which can help to keep weight off in the long term.
3. Mental health benefits: Physical activity releases endorphins, which are natural mood elevators, reducing stress, anxiety, and depression. Regular exercise can also improve sleep quality, increase self-esteem, and provide a sense of accomplishment, contributing to overall mental well-being.
4. Enhanced immune function: Regular exercise can boost the immune system, making the body more effective at fighting off infections and diseases. It can also reduce inflammation, which is linked to many chronic diseases.
5. Increased bone density: Weight-bearing exercises like walking, running, and strength training can help to increase bone density, reducing the risk of osteoporosis and fractures.
6. Better flexibility and mobility: Exercise can improve flexibility, balance, and coordination, reducing the risk of falls and injuries. It can also improve posture and range of motion, leading to better overall physical performance.
7. Enhanced cognitive function: Regular physical activity has been linked to improved brain function, including better memory, focus, and cognitive abilities. It can also reduce the risk of cognitive decline and diseases like Alzheimer's.
8. Stronger muscles and joints: Exercise helps to build and maintain muscle mass, which supports joint health and reduces the risk of injury. It also improves overall body strength and stability, making daily activities easier and less strenuous.
9. Better sleep: Regular exercise can help regulate sleep patterns, leading to deeper, more restful sleep. It can also reduce the risk of sleep disorders and improve overall sleep quality.
10. Social benefits: Exercise can provide opportunities for social interaction and community engagement, leading to improved social connections and a sense of belonging.
11. Reduced risk of chronic diseases: Regular exercise has been shown to reduce the risk of many chronic diseases, including type 2 diabetes, certain types of cancer, and neurodegenerative diseases like Alzheimer's.
12. Longevity: Studies have shown that regular exercise can increase lifespan and improve overall quality of life, as it helps to maintain physical and mental health, reduces the risk of chronic diseases, and promotes healthy aging.
13. Enhanced energy levels: Exercise can increase energy levels and stamina, making it easier to perform daily activities and maintain an active lifestyle.
14. Stress relief: Physical activity can help to reduce stress and anxiety, providing a natural and healthy way to manage and cope with life's challenges.
15. Improved digestion: Exercise can help to improve digestion and prevent constipation, as well as reduce the risk of digestive disorders like irritable bowel syndrome and colon cancer.
16. Better skin health: Exercise promotes healthy blood flow, which can improve skin health and reduce the appearance of wrinkles and fine lines. It can also help to detoxify the body, leading to clearer, healthier skin.
17. Increased independence: Regular exercise can improve balance, strength, and mobility, leading to increased independence and a better quality of life, especially for older adults.
18. Enhanced creativity: Physical activity has been linked to improved creativity and problem-solving abilities, as it stimulates the brain and promotes new connections between neurons.
19. Better time management: Exercise can help to improve time management and productivity, as it provides a healthy outlet for stress and anxiety, and promotes a sense of accomplishment and well-being.
20. Environmental benefits: Exercise can reduce the carbon footprint by promoting active transportation like walking or biking, reducing the need for cars and contributing to a cleaner, healthier environment.
Overall, regular exercise has numerous benefits for physical, mental, and emotional health, as well as social, environmental, and economic benefits. It is a crucial component of a healthy and fulfilling lifestyle, and should be a priority for everyone.
```
</details>

**Comment by b-bargabus** [comment]
> > Even the very simple test that o7si used failed badly on instruction following producing just two results when a clear prompt asked for three.
> 
> Small clarification, that was just `-n 120` truncating the output. I re-ran it without the `-n` cap:
> 

OK, much larger instruction following collapse ;-)

**Comment by o7si** [comment]
@ggerganov Just to confirm the direction for this PR:

1. **Warning only (as you suggested):** the code is ready on a separate branch (872bc76) and I can force-push it here anytime.
2. **Keep the current fix:** slightly more complex, but it avoids the quality regression for short inputs at large `-c` (e.g. f1 dropped from 0.95 to 0.35 in @b-bargabus's evals).

If you think option 2 isn't worth the extra complexity, I'll push the warning-only version. Thanks!
